src/audio/mux/mux.c

```diff
@@ -200,7 +205,230 @@ static struct comp_dev *mux_new(const struct comp_driver *drv,
 	dev->state = COMP_STATE_READY;
 	return dev;
 }
+#else
+static int build_config(struct comp_data *cd, struct comp_dev *dev)
+{
+	dev->ipc_config.type = SOF_COMP_MUX;
+	cd->config.frame_format_deprecated = cd->md.base_cfg.audio_fmt.depth;
+	cd->config.num_streams = 2;
+	int mask = 1;
+
+	/* clear masks */
+	for (uint16_t i = 0; i < cd->config.num_streams; i++)
+		for (uint16_t j = 0 ; j < PLATFORM_MAX_CHANNELS; j++)
+			cd->config.streams[i].mask[j] = 0;
+	/*
+	 * Setting masks for streams. Channels for IPC4 are mapped
+	 * according to pin and channel order. First the first input stream, then the
+	 * reference. For example, for base config 2-channel and reference 2-channel, masks
+	 * should look like: mask[] = { 0b00000001, 0b00000010 } for the first
+	 * stream and mask[] = { 0b00000100, 0b00001000 } for the second
+	 * (reference)
+	 *           +---+           +---+
+	 *           | 0 |---------> | 0 |
+	 * INPUT     +---+           +---+
+	 * STREAM 0  | 1 |---------> | 1 |
+	 *           +---+           +---+  OUTPUT
+	 *                    +----> | 2 |  STREAM
+	 *           +---+    |      +---+
+	 *           | 0 |----+  +-> | 3 |
+	 * INPUT     +---+       |   +---+
+	 * STREAM 1  | 1 |-------+
+	 *           +---+
+	 */
+	for (uint16_t i = 0; i < cd->md.base_cfg.audio_fmt.channels_count; i++) {
+		cd->config.streams[0].mask[i] = mask;
+		mask = mask * 2;
+	}
+
+	for (uint16_t i = 0; i < cd->md.reference_format.channels_count; i++) {
+		cd->config.streams[1].mask[i] = mask;
+		mask = mask * 2;
+	}
+
+	/* verification that masks are correct */
+	for (uint16_t i = 0; i < cd->config.num_streams; i++) {
+		for (uint16_t j = 0 ; j < PLATFORM_MAX_CHANNELS; j++) {
+			if (popcount(cd->config.streams[i].mask[j]) > 1) {
```

Comment:
wait, you just calculated them, right? Isn't it similar to
```
x = 2 + 2;
if (x != 4) return -ECPUMELTDOWN;
```
? Or can there be more than 2 streams so that the second and any further streams weren't set up above? And you probably want to `memset(0)` the whole array at the top of the function if it's important that skipped masks are valid too?

Reply:
but how can any of those masks contain more than 1 set bit? You begin setting them with `mask = 1` and then just shift the mask to the left, so each mask, that has been set, will have exactly one bit set.

Reply:
I removed the redundant verification and added memset